kolibri_explore_plugin/collectionviews.py

```diff
@@ -0,0 +1,621 @@
+            "current_task_number": current_task_number,
+            "total_tasks_number": total_tasks_number,
+            "extra_metadata": extra_metadata,
+        }
+
+    def to_state(self):
+        """Create and return a state representation that is JSON serializable.
+
+        The state can be persisted and used in another session to
+        regenerate this singleton using the from_state() method.
+        """
+        grade = None
+        name = None
+        if self._content_manifest is not None:
+            grade = self._content_manifest.grade
+            name = self._content_manifest.name
+
+        state = {
+            "grade": grade,
+            "name": name,
+            "stage": self._stage.name,
+            "current_job_id": self._current_job_id,
+            "current_task": self._current_task,
+            "tasks_pending": self._tasks_pending,
+            "tasks_completed": self._tasks_completed,
+            "tasks_previously_completed": self._tasks_previously_completed,
+        }
+
+        return state
+
+    def _next_task_or_stage(self, user):
+        if not self._tasks_pending:
+            # No more tasks pending in this stage, move to the next one
+            self._set_next_stage()
+        else:
+            # Enqueue next pending task
+            self._enqueue_current_task(user)
+
+    def _set_next_stage(self):
+        if self._stage == DownloadStage.COMPLETED:
+            return
+
+        self._stage = DownloadStage(self._stage + 1)
+
+        tasks = []
+        if self._stage == DownloadStage.IMPORTING_CHANNELS:
+            tasks = self._content_manifest.get_channelimport_tasks()
+        elif self._stage == DownloadStage.IMPORTING_CONTENT:
+            tasks = self._content_manifest.get_contentimport_tasks()
+        self._tasks_pending = tasks
+        self._tasks_previously_completed.extend(self._tasks_completed)
+        self._tasks_completed = []
+
+    def _enqueue_current_task(self, user):
+        self._current_task = self._tasks_pending.pop(0)
+
+        self._enqueue_timestamp = time.time()
+        logger.debug(
+            f"Enqueuing task {self._current_task}"
+            f" at {self._enqueue_timestamp}"
+        )
+
+        tasks_mapping = {
+            "remotechannelimport": _remotechannelimport,
+            "remotecontentimport": _remotecontentimport,
+        }
+        fn = tasks_mapping[self._current_task["task"]]
+        params = self._current_task["params"]
+        self._current_job_id = fn(user=user, **params)
+        self._enqueue_timestamp = None
+        logger.debug(f"Enqueued job id {self._current_job_id}")
+
+
+def _remotechannelimport(user, channel_id):
+    """Create, validate and enqueue a remotechannelimport job."""
+    job = remotechannelimport.validate_job_data(
+        user,
+        {
+            "channel_id": channel_id,
+            # FIXME: Why is channel_name needed? It isn't known at this point.
+            "channel_name": "foo",
```

Comment:
That happened in this pull request:

https://github.com/learningequality/kolibri/pull/9438/commits/b53d7baa29374b4dbfb467573b664c9175bfb958#diff-0ebb74f1619d471f9ba176a2414e427e279a76fac0278f71fda6414a441ab0b5L109-L111

Looks like it used to set a default, but it was unused so it got removed. That old default looked like this:

```
def get_channel_name(channel_id, require_channel=False):
    try:
        channel = ChannelMetadata.objects.get(id=channel_id)
        channel_name = channel.name
    except ChannelMetadata.DoesNotExist:
        if require_channel:
            raise serializers.ValidationError("This channel does not exist")
        channel_name = ""

    return channel_name
```

Does the job validator accept an empty string for channel_name? That would be the same as the behaviour with Kolibri 0.15.

Reply:
> Does the job validator accept an empty string for channel_name? That would be the same as the behaviour with Kolibri 0.15.

Nope, if I pass an empty string the validation fails with a "channel name can't be empty" error.